src/blas/impl/KokkosBlas_Newton_impl.hpp

```diff
@@ -0,0 +1,250 @@
+
+  System sys;
+  XVector x;
+  YVector rhs;
+  NewtonHandleType handle;
+  int N;
+
+  Matrix J, tmp;
+  XVector update;
+  Kokkos::View<norm_type, execution_space> norm;
+  Kokkos::View<yvalue_type*, execution_space> alpha;
+
+  NewtonFunctor(System _sys, XVector _x, YVector _rhs,
+                NewtonHandleType& _handle)
+      : sys(_sys), x(_x), rhs(_rhs), handle(_handle) {
+    J      = Matrix("Jacobian", x.extent(0), x.extent(0));
+    tmp    = Matrix("Jacobian", x.extent(0), x.extent(0) + 4);
+    update = XVector("update", x.extent(0));
+    norm   = Kokkos::View<norm_type, execution_space>("Newton norm");
+    alpha  = Kokkos::View<yvalue_type*, execution_space>("alpha", 1);
+    N      = x.extent(0);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  NewtonSolverStatus solve() const {
+    alpha(0)               = Kokkos::ArithTraits<yvalue_type>::one();
+    handle.lastResidual(0) = -1;  // init to dummy value
+
+    // Iterate until maxIts or the tolerance is reached
+    for (int it = 0; it < handle.maxIters; ++it) {
+      // compute initial rhs
+      sys.derivatives(0, x, rhs);
+      if (handle.debug_mode) {
+        KOKKOS_IMPL_DO_NOT_USE_PRINTF("NewtonFunctor: r=");
+        for (int k = 0; k < N; k++) {
+          KOKKOS_IMPL_DO_NOT_USE_PRINTF("%f \n", rhs(k));
+        }
+      }
+
+      // Solve the following linearized
+      // problem at each step: J*update=-rhs
+      // with J=du/dx, rhs=f(u_n+update)-f(u_n)
+      int ierr = KokkosBlas::serial_nrm2(rhs, norm);
+      if (0 < ierr) {
+        KOKKOS_IMPL_DO_NOT_USE_PRINTF(
+            "NewtonFunctor: serial_nrm2 reports the following error %d\n",
+            ierr);
+      }
+      handle.lastResidual(0) = norm();
+
+      if (handle.debug_mode) {
+        KOKKOS_IMPL_DO_NOT_USE_PRINTF(
+            "NewtonFunctor: Iteration: %d  Current res norm is: %e \n Current "
+            "soln is:\n",
+            // it, norm(0));
+            it, (double)handle.lastResidual(0));
+        for (int k = 0; k < N; k++) {
+          KOKKOS_IMPL_DO_NOT_USE_PRINTF("%f \n", x(k));
+        }
+      }
+
+      if (norm() < handle.relativeTol) {
+        // Problem solved, exit the functor
+        if (handle.debug_mode) {
+          KOKKOS_IMPL_DO_NOT_USE_PRINTF(
+              "NewtonFunctor: Newton solver converged! Ending norm is: %e \n "
+              "Solution x is: "
+              "\n",
+              norm());
+          for (int k = 0; k < N; k++) {
+            KOKKOS_IMPL_DO_NOT_USE_PRINTF("%f \n", x(k));
+          }
+        }
+        return NewtonSolverStatus::Converged;
+      }
+
+      // compute LHS
+      sys.jacobian(0, x, J);
+
+      // solve linear problem
+      int linSolverStat = KokkosBatched::SerialGesv<
+          KokkosBatched::Gesv::StaticPivoting>::invoke(J, update, rhs, tmp);
+      KokkosBlas::SerialScale::invoke(-1, update);
+
+      // update state of variables if needed
+      // sys.update_state(update);
```

Comment:
Why is this commented out?

Reply:
That might need to be brought back in, I am not too sure on whether we need this function or not?